src/constitutive/TACSSmearedCompositeShellConstitutive.cpp

```diff
@@ -0,0 +1,600 @@
+    int elemIndex, const double pt[], const TacsScalar X[],
+    const TacsScalar strain[]) {
+  int nvals = 2 * num_plies;
+  TacsScalar *fvals = new TacsScalar[nvals];
+
+  evalPlyTopBottomFailure(strain, fvals);
+  TacsScalar ks_val = ksAggregation(fvals, nvals, ks_weight);
+
+  delete[] fvals;
+
+  return ks_val;
+}
+
+/*
+  Compute the most critical failure criteria for the laminate
+*/
+void TACSSmearedCompositeShellConstitutive::evalPlyTopBottomFailure(
+    const TacsScalar strain[], TacsScalar fvals[]) {
+  // Compute the total thickness of the laminate
+  TacsScalar tb = -0.5 * thickness;
+  TacsScalar tt = 0.5 * thickness;
+
+  for (int i = 0; i < num_plies; i++) {
+    TacsScalar lamStrain[3] = {0.0};
+
+    getLaminaStrain(strain, tb, lamStrain);
+    fvals[2 * i] = ply_props[i]->failure(ply_angles[i], lamStrain);
+
+    getLaminaStrain(strain, tt, lamStrain);
+    fvals[2 * i + 1] = ply_props[i]->failure(ply_angles[i], lamStrain);
+  }
+}
+
+// Evaluate the derivative of the failure criteria w.r.t. the strain
+TacsScalar TACSSmearedCompositeShellConstitutive::evalFailureStrainSens(
+    int elemIndex, const double pt[], const TacsScalar X[],
+    const TacsScalar strain[], TacsScalar sens[]) {
+  int nvals = 2 * num_plies;
+  int nvars = 9;
+
+  sens[0] = sens[1] = sens[2] = 0.0;
+  sens[3] = sens[4] = sens[5] = 0.0;
+  sens[6] = sens[7] = sens[8] = 0.0;
+
+  // Compute the total thickness of the laminate
+  TacsScalar tb = -0.5 * thickness;
+  TacsScalar tt = 0.5 * thickness;
+
+  TacsScalar *fvals = new TacsScalar[nvals];
+  evalPlyTopBottomFailure(strain, fvals);
+
+  TacsScalar **dfvals = new TacsScalar *[nvals];
+  for (int i = 0; i < num_plies; i++) {
+    TacsScalar lamStrain[3], phi[3];
+
+    // plate bottom stress
+    dfvals[2 * i + 0] = new TacsScalar[nvars];
+    getLaminaStrain(strain, tb, lamStrain);
+    TacsScalar fval_bot =
+        ply_props[i]->failureStrainSens(ply_angles[i], lamStrain, phi);
+    dfvals[2 * i + 0][0] = phi[0];
+    dfvals[2 * i + 0][1] = phi[1];
+    dfvals[2 * i + 0][2] = phi[2];
+    dfvals[2 * i + 0][3] = tb * phi[0];
+    dfvals[2 * i + 0][4] = tb * phi[1];
+    dfvals[2 * i + 0][5] = tb * phi[2];
+    dfvals[2 * i + 0][6] = dfvals[2 * i + 0][7] = dfvals[2 * i + 0][8] = 0.0;
+
+    // plate top stress
+    dfvals[2 * i + 1] = new TacsScalar[nvars];
+    getLaminaStrain(strain, tt, lamStrain);
+    TacsScalar fval_top =
+        ply_props[i]->failureStrainSens(ply_angles[i], lamStrain, phi);
+    dfvals[2 * i + 1][0] = phi[0];
+    dfvals[2 * i + 1][1] = phi[1];
+    dfvals[2 * i + 1][2] = phi[2];
+    dfvals[2 * i + 1][3] = tt * phi[0];
+    dfvals[2 * i + 1][4] = tt * phi[1];
+    dfvals[2 * i + 1][5] = tt * phi[2];
+    dfvals[2 * i + 1][6] = dfvals[2 * i + 1][7] = dfvals[2 * i + 1][8] = 0.0;
+  }
+
+  TacsScalar ks_val =
+      ksAggregationSensProduct(fvals, nvals, nvars, ks_weight, dfvals, sens);
+
+  for (int i = 0; i < nvals; i++) {
+    delete[] dfvals[i];
+  }
+
+  delete[] fvals;
```

Comment:
It might not matter since these partials are generally much cheaper than the adjoint solve, but I think allocating and deleting memory like this inside functions that are going to be called inside "hot" loops is not good for performance. In the `BladeStiffenedShell` class I got around this by making these arrays attributes of the class that get created/destroyed at the same time as the constitutive object

Reply:
I've moved all of these arrays to the construction/destruction method for the class